### PR TITLE
Improve layout configuration part of the user guide

### PR DESCRIPTION
**Pull request by pcahyna**
##### Pull Request Details:

* Type: **Enhancement**

* Impact: **Low**

* Reference to related issue (URL): related to #2997

* How was this pull request tested?
Not tested.

* Description of the changes in this pull request:
Improvements of the layout configuration part of the user guide to address #2997
  *   Describe `EXCLUDE_RECREATE` better. Instead of copying a snippet of default.conf, describe in prose what the variable does and what is the syntax.
  *  Update whitespace after comment in layout examples. Commit 4d6fac780f9234b55dbb0ca7fa55b56e4fbc7157 changed the formatting of commented entries in layout: "Change commenting in layout file to have no space after the #." Adapt the examples that had been generated before the change to match.
  *  Do not copy almost identical layout file in the documentation multiple times. Show only the parts that have changed after changing something.
  * Move autoexcludes description. I believe that "This behavior is controlled by the `AUTOEXCLUDE_DISKS=y` parameter" must refer to the exclusion of the unused backup disk. At the current location (disk layout when the backup disk is NOT unused anymore) it does not make sense. Move the whole part about autoexcludes just after the section where it is mentioned that a disk was automatically excluded.
  * Don't copy EXCLUDE_BACKUP/RESTORE from default.conf. `EXCLUDE_RESTORE` and `EXCLUDE_BACKUP` are irrelevant for layout file and disk layout recreation. They merely influence file backup or restore (for supported backup methods). This guide is about the recreation of the disk layout. Therefore, omit the variables from here. They are not used in the doc anywhere anyway after showing a copy of their description originating from default.conf. So, why to show it at all?
  * Do not refer to "restore list". This term is never defined or used anywhere. Let's talk about the layout file instead.

Open question: where to describe variables that control only the file backup and restoration (for supported backup methods)? Like `EXCLUDE_RESTORE` and `EXCLUDE_BACKUP` and `BACKUP_PROG_EXCLUDE`. They don't belong here as this chapter is only about storage layout recreation. I think it would be the best to introduce chapter 7 with this information.

### CODE REVIEW

**Comment by pcahyna** [comment]
@danboid can you please have a look whether this helps with your issue #2997 ?

**Comment by pcahyna** [comment]
diff gets confused by the move of a large piece of text, so the diff of the whole PR does not make much sense. I suggest to look at per-commit diffs (https://github.com/rear/rear/pull/3125/commits) when reviewing the PR, each of them looks reasonable.

**Comment by danboid** [comment]
Thanks for working on the rear docs!

I've had a look at these commits but I have nothing to add really.

I might be able to comment better once committed and I ca read it as a wholebut even then, I don't know rear well enough to be correcting one of the devs.

**Comment by jsmeix** [comment]
@danboid FYI:
you can see the changes via
https://github.com/rear/rear/pull/3125/files
(it even works when one is not logged in at GitHub)
and therein at the top right corner
of the sub-window that shows the changes for a file
click on the three dots '...' which shows a sub-menu
where you can select "View file"
so you can read the changed file as a whole.

File: doc/user-guide/06-layout-configuration.adoc [hunk 1]
```diff
@@ -143,61 +106,60 @@ we mount the +/dev/mapper/backup-backup+ filesystem on
 +/media/backup+, it will get excluded, as the mountpoint is under
 +/media+.
 
-=== Manual excludes
-It seems prudent to prevent the external drives from ever being backed-up or overwritten. The default configuration contains these lines:
+If we mount the filesystem on +/dev/mapper/backup-backup+ on +/backup+,
+Relax-and-Recover will think that it's necessary to recreate the filesystem:
```
Comment (jsmeix) [inline doc/user-guide/06-layout-configuration.adoc:110]: Could you rephrase it so that it does no longer look
as if Relax-and-Recover has the ability to think
e.g. like
```
If we mount the filesystem on /dev/mapper/backup-backup on /backup,
Relax-and-Recover will recreate this filesystem
Relax-and-Recover recreates mounted filesystems
except filesystems that get excluded by default.
```
This is only an offhanded proposal.
I did not verify that ReaR really recreates filesystems
only depending on whether or not they are mounted.

File: doc/user-guide/06-layout-configuration.adoc [hunk 2]
```diff
@@ -143,61 +106,60 @@ we mount the +/dev/mapper/backup-backup+ filesystem on
...
+...
+lvmvol /dev/backup backup 12800 104857600 
+lvmvol /dev/backup externaltemp 38400 314572800 
+...
+fs /dev/mapper/backup-backup /backup ext4 uuid=da20354a-dc4c-4bef-817c-1c92894bb002 label= blocksize=4096 reserved_blocks=655360 max_mounts=24 check_interval=180d options=rw
 ----------------------------------
 
-To prevent an inadvertently mounted backup filesystem being added to the restore list, the easiest way is to add the filesystem to the +EXCLUDE_RECREATE+ array.
+=== Manual excludes
+It seems prudent to prevent the external drives from ever being backed-up or overwritten.
```
Comment (jsmeix) [inline doc/user-guide/06-layout-configuration.adoc:127]: "It seems prudent" sounds smart-alecky.
Simpler and more neutral wording could be like
```
Normally data on external drives should
neither be backed-up during "rear mkbackup"
nor restored during "rear recover".
```

By the way:
I find this whole part confusing.
For me "backup and restore" means backup and restore of data
(files and directories - i.e. what there is in backup.tar.gz)
but here it looks as if disk layout is meant because it then
talks about EXCLUDE_RECREATE which is primarily to exclude
disk layout components during disk layout recreation
and adding them also to EXCLUDE_BACKUP is only an
automated side effect.

File: doc/user-guide/06-layout-configuration.adoc [hunk 3]
```diff
@@ -143,61 +106,60 @@ we mount the +/dev/mapper/backup-backup+ filesystem on
...
+specify a mounted filesystem, use the +fs:+ prefix:
++fs:+<mountpoint>. To specify swap, use the +swap:+ prefix. The syntax
+to specify a volume group to exclude (together with all its logical
+volumes and filesystems on them) is +/dev/+<volume group name>. Consult
+the +/var/lib/rear/layout/disktodo.conf+ file created together with
+the layout file for the full list of components in a compatible
+syntax.
+
+To prevent an inadvertently mounted backup filesystem being added to
+the layout file and recreated, the easiest way is to add the filesystem to the +EXCLUDE_RECREATE+ array.
```
Comment (jsmeix) [inline doc/user-guide/06-layout-configuration.adoc:144]: Ouch! Argh! Ugh! ;-)

Please remove all
* just
* simple
* easy

words from all documentation everywhere.

Those words are never helpful for a user.
Those words make a desperate user who reads documentation angry
because those words humiliate the user because they basically
tell him he is somewhat stupid when he has issues with things
that are just simple and easy for others.

It is in particular humiliating and patronizing
when the developers tell their users
that things are just simple and easy.

Users who read documentation are often desperate
because reading documentation is often the last thing
one does when (easy) things (simply) do not (just) work.

File: doc/user-guide/06-layout-configuration.adoc (hunk 2 above)
Comment (pcahyna) [reply doc/user-guide/06-layout-configuration.adoc:127]: > For me "backup and restore" means backup and restore of data (files and directories - i.e. what there is in backup.tar.gz) but here it looks as if disk layout is meant because it then talks about EXCLUDE_RECREATE which is primarily to exclude disk layout components during disk layout recreation and adding them also to EXCLUDE_BACKUP is only an automated side effect.

The point of this document is indeed disk layout save and recreation, but I think this phrase is intended to mean that one wants to exclude the drive both from data backup and layout. After all, it is not wrong to point out that `EXCLUDE_RECREATE` does both. I will think how to rephrase it in order to make clear that the main point is exclusion from layout, with exclusion from backup only "by the way".

File: doc/user-guide/06-layout-configuration.adoc [hunk 4]
```diff
@@ -143,61 +106,60 @@ we mount the +/dev/mapper/backup-backup+ filesystem on
...
+lvmvol /dev/backup backup 12800 104857600 
+lvmvol /dev/backup externaltemp 38400 314572800 
+...
+fs /dev/mapper/backup-backup /backup ext4 uuid=da20354a-dc4c-4bef-817c-1c92894bb002 label= blocksize=4096 reserved_blocks=655360 max_mounts=24 check_interval=180d options=rw
 ----------------------------------
 
-To prevent an inadvertently mounted backup filesystem being added to the restore list, the easiest way is to add the filesystem to the +EXCLUDE_RECREATE+ array.
+=== Manual excludes
+It seems prudent to prevent the external drives from ever being backed-up or overwritten.
+The most generic mechanism for doing this is the +EXCLUDE_RECREATE+
```
[the remark this answers is not in the document]
Comment (pcahyna) [reply doc/user-guide/06-layout-configuration.adoc:128]: > why a superlative judgement here?

because there are less generic inclusion/exclusion rules in ReaR, like `EXCLUDE_VG` or `ONLY_INCLUDE_VG`, and `EXCLUDE_RECREATE` is the most generic of them (can exclude anything).

File: doc/user-guide/06-layout-configuration.adoc (hunk 3 above)
Comment (pcahyna) [reply doc/user-guide/06-layout-configuration.adoc:144]: @jsmeix I actually share your opinion on the general tone and style of this whole document (applies to most of your review comments), but it was not my goal to rewrite it entirely, only to improve its factual correctness and logical structure.

If you review the commits one by one, as suggested in https://github.com/rear/rear/pull/3125#issuecomment-1883420901 , you will see that most of the wording that you don't like has been there since the beginning.

File: doc/user-guide/06-layout-configuration.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (jsmeix) [reply doc/user-guide/06-layout-configuration.adoc:128]: Is there something that is even more specific as EXCLUDE_VG?
So we would have
* EXCLUDE_RECREATE is most generic
* EXCLUDE_VG is generic
* ??? is specific

This is perhaps a stupid question
but I really have understanding problems
with all that include/exclude stuff in ReaR.

File: doc/user-guide/06-layout-configuration.adoc (hunk 3 above)
Comment (jsmeix) [reply doc/user-guide/06-layout-configuration.adoc:144]: I know well that most of the wording I don't like
is there since the beginning.

So no need to worry now about the general tone and style.

Perhaps later I may (provided I find time for it)
clean up the general tone and style.

**Review by jsmeix — APPROVED** [review]
@pcahyna 
thank you for proof-reading our documentation
and for your fixes!

**Review by schlomo — APPROVED** [review]
Thank you @pcahyna for improving the documentation, any clean-as-you-go is of course also appreciated.

@jsmeix I think that the main point is an improvement and let's be happy with whatever we can get as documentation contributions are already rather seldom.

And yes, I agree that we should try to avoid overly "fluffy" language. And we have a lot of old and older stuff in ReaR, where we won't have time to fix it all any time soon.

File: doc/user-guide/06-layout-configuration.adoc (hunk 1 above)
Comment (pcahyna) [reply doc/user-guide/06-layout-configuration.adoc:110]: > I did not verify that ReaR really recreates filesystems
> only depending on whether or not they are mounted.

I verified that, actually.

File: doc/user-guide/06-layout-configuration.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (pcahyna) [reply doc/user-guide/06-layout-configuration.adoc:128]: @jsmeix I don't think so. There is `EXCLUDE_MOUNTPOINTS`, which one could consider either more or less specific as `EXCLUDE_VG` as its scope is different.
The point of the superlative is that there is nothing more generic than `EXCLUDE_RECREATE` (which can include any component), but perhaps even that is not literally correct, as `AUTOEXCLUDE_PATH` could be considered more genrric in some sense (can exclude only filesystems, but their mountpoints can be specified as a directory prefix).

[the remark this answers is not in the document]
Comment (pcahyna) [reply doc/user-guide/06-layout-configuration.adoc:128]: @jsmeix addressed in bfad3f3207fefa8766fe0866ed21ca50a9b604b8

File: doc/user-guide/06-layout-configuration.adoc [hunk 5]
```diff
@@ -143,61 +106,60 @@ we mount the +/dev/mapper/backup-backup+ filesystem on
...
+a component in the layout file, but the syntax is different. To
+specify a mounted filesystem, use the +fs:+ prefix:
++fs:+<mountpoint>. To specify swap, use the +swap:+ prefix. The syntax
+to specify a volume group to exclude (together with all its logical
+volumes and filesystems on them) is +/dev/+<volume group name>. Consult
+the +/var/lib/rear/layout/disktodo.conf+ file created together with
+the layout file for the full list of components in a compatible
+syntax.
+
+To prevent an inadvertently mounted backup filesystem being added to
```
[the remark this answers is not in the document]
Comment (pcahyna) [reply doc/user-guide/06-layout-configuration.adoc]: > Again I wonder why "inadvertently" is needed here.
> Why judgement about why that filesystem is mounted?
> Perhaps it is mounted intentionally?

@jsmeix addressed in d1440889db7d162e47bc6cfa37e23709150c5456

[the remark this answers is not in the document]
Comment (pcahyna) [reply doc/user-guide/06-layout-configuration.adoc]: > Also why only prevent a "backup" filesystem?

that's the assumption of the example. I tried to describe that in a899b9677b7c0f2e70ea62df4665de911c6f2416

File: doc/user-guide/06-layout-configuration.adoc (hunk 3 above)
Comment (pcahyna) [reply doc/user-guide/06-layout-configuration.adoc]: Partly addressed in d1440889db7d162e47bc6cfa37e23709150c5456 (for the parts that I have touched).

**Comment by pcahyna** [comment]
Hi @jsmeix @schlomo thank you for looking, I made some changes according to your comments.

Any opinion on
> Open question: where to describe variables that control only the file backup and restoration (for supported backup methods)? Like EXCLUDE_RESTORE and EXCLUDE_BACKUP and BACKUP_PROG_EXCLUDE. They don't belong here as this chapter is only about storage layout recreation. I think it would be the best to introduce chapter 7 with this information.

?

File: doc/user-guide/06-layout-configuration.adoc (hunk 3 above)
Comment (jsmeix) [reply doc/user-guide/06-layout-configuration.adoc]: For the future I filed
https://github.com/rear/rear/issues/3129

File: doc/user-guide/06-layout-configuration.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsmeix) [reply doc/user-guide/06-layout-configuration.adoc]: See also https://github.com/rear/rear/issues/3129

File: doc/user-guide/06-layout-configuration.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (jsmeix) [reply doc/user-guide/06-layout-configuration.adoc]: See also https://github.com/rear/rear/issues/3129

File: doc/user-guide/06-layout-configuration.adoc (hunk 2 above)
Comment (jsmeix) [reply doc/user-guide/06-layout-configuration.adoc]: See also https://github.com/rear/rear/issues/3129

File: doc/user-guide/06-layout-configuration.adoc (hunk 1 above)
Comment (jsmeix) [reply doc/user-guide/06-layout-configuration.adoc:109]: See also https://github.com/rear/rear/issues/3129

**Comment by jsmeix** [comment]
I think a new chapter that only describes variables
that control only the file backup and restore
is not possible in practice because of the various
interdependencies with the variables that control
the disk layout.

Right now I see that we have already a chapter 7
https://github.com/rear/rear/blob/master/doc/user-guide/07-tips-and-tricks.adoc
which I had never read before.

**Comment by pcahyna** [comment]
Yes, unfortunately the chapters are numbered, so to insert a new one we would have to renumber the following ones.

@jsmeix do you think the description of file backup-related variables should be added also to this chapter? This would make it less focused, though.

**Comment by jsmeix** [comment]
@pcahyna 
offhandedly I think disk layout and files backup
should be kept separated primarily because
```
Relax-and-Recover (ReaR) complements backup and restore
of files but ReaR is neither a backup software
nor a backup management software
and it is not meant to be one.

In general backup and restore of the files
is external functionality for ReaR.
```
https://en.opensuse.org/SDB:Disaster_Recovery#Relax-and-Recover_versus_backup_and_restore

**Comment by jsmeix** [comment]
At first glance I think
https://github.com/rear/rear/blob/master/doc/user-guide/07-tips-and-tricks.adoc
is not really useful - perhaps even bad in some cases, for example:

* I think GRUB_RESCUE can be useful but at the same time
it is a bad idea to let ReaR modify the bootloader of the
original system
* Why it is a "tip" or a "trick" that it automatically detects
and enables serial console support?
* Preserving SSH keys can be useful but at the same time
it is a bad idea to leak out secret SSH keys from the original system
into the (possibly "public" accessible) recovery system (initrd/ISO/...)
* During recovery one can NOT at any stage re-run "rear recover". In my experience in most cases re-running "rear recover" fails or would only work with too complicated manual things so that in most cases it works better in practice to reboot the recovery system, do needed adaptions, and then run "rear recover".

So from my current point of view
https://github.com/rear/rear/blob/master/doc/user-guide/07-tips-and-tricks.adoc
is mostly questionable / misleading / wrong / ...

**Comment by pcahyna** [comment]
> * Preserving SSH keys can be useful but at the same time
>       it is a bad idea to leak out secret SSH keys from the original system
>       into the (possibly "public" accessible) recovery system (initrd/ISO/...)

The document does not suggest to leak secrets. I believe that what is intended is to mention the preservation of root's `.ssh/authorized_keys`, which allows you to login remotely. This file is not especially sensitive.

**Comment by pcahyna** [comment]
> * During recovery one can NOT at any stage re-run "rear recover".
>       In my experience in most cases re-running "rear recover" fails
>       or would only work with too complicated manual things
>       so that in most cases it works better in practice
>       to reboot the recovery system, do needed adaptions,
>       and then run "rear recover" again.

Maybe. But the whole layout recreation is implemented with this in mind, with all the `todo` and `done` in disktodo.conf. This is also the point of the sentence just above https://github.com/rear/rear/blob/master/doc/user-guide/06-layout-configuration.adoc#planning-in-advance :

> Notice how Relax-and-Recover detected that it had already created quite a few components and did not try to recreate them anymore.

I must say I don't like the whole section https://github.com/rear/rear/blob/master/doc/user-guide/06-layout-configuration.adoc#restore-to-different-hardware , WDYT?

**Comment by pcahyna** [comment]
@danboid can you please have a look at the resulting document https://github.com/rear/rear/blob/16d6be5664a24ce5f30e5f8b9cb727e89a15c37f/doc/user-guide/06-layout-configuration.adoc ? Not to check correctness, but to see whether it is better understandable than the previous version and addresses https://github.com/rear/rear/issues/2997 . (Note that the backup options EXCLUDE_BACKUP etc. are deliberately omitted, as they don't refer to the layout configuration.)

**Comment by jsmeix** [comment]
@pcahyna 
the section
https://github.com/rear/rear/blob/master/doc/user-guide/06-layout-configuration.adoc#restore-to-different-hardware
is rather old, in particular I never updated it.

I did not pay attention to this section because
at SUSE we do not support migration to different hardware.

In general I find migration to different hardware
not at all something that is "simple" with ReaR.

I think what that section basically tells is that
migration to different hardware does not "just work".

So the basic information is still true and users
can clearly see that with an example where the technical
details are somewhat outdated which is a minor issue here
because nobody has exactly such a system so the details
should not matter in practice for this section.

**Comment by pcahyna** [comment]
@jsmeix I especially dislike the wording
> There are two ways to deal with different hardware. One is being lazy and dealing with problems when you encounter them. The second option is to plan in advance. Both are valid approaches.

as it sounds like one can leave figuring out the details to the moment when disaster happens (although this may not have been the original intent of the text).

**Comment by jsmeix** [comment]
@pcahyna 
according to
```
# git log --follow -p doc/user-guide/06-layout-configuration.adoc
```
that part about "being lazy" is unchanged
since it was introduced at the beginning as
"a first draft of a layout engine usage guide" via
https://github.com/rear/rear/commit/259b66cd59a0cde9cf0070c2fa3048c7ddb13664

I think what it tries to tell is that
* the lazy approach is meant when the target hardware
is unknown in advance and then one must have
good knowledge of one's system - actually the lazy approach
is the only possible way in practice when one does not
already have replacement hardware to verify it in advance
(e.g. for individual end user workstations and laptops)
* the plan in advance approach is meant ("is preferable")
when the target hardware is known in advance
(in particular for business/enterprise environments)

At SUSE we officially support only the latter case, cf.
https://en.opensuse.org/SDB:Disaster_Recovery and
https://documentation.suse.com/sle-ha/15-SP5/html/SLE-HA-all/cha-ha-rear.html

**Comment by jsmeix** [comment]
@pcahyna 
is it good enough so that it can be merged
(even if some parts are not yet perfectly well)
or are further improvements required in this pull request?

**Comment by pcahyna** [comment]
@jsmeix I have been hoping that my professional doc writer colleague would review the PR and possibly even help with the numerous formal/style problems that we identified, that's why I have kept the PR open. Does it conflict with your work?

**Comment by jsmeix** [comment]
@pcahyna 
no problem, take your time.
There is no conflict with my work.
I was only wondering why it did not proceed.

**Comment by gdha** [comment]
@pcahyna @jsmeix Moved the milestone to "3.0" - hope that is ok for you?

**Comment by jsmeix** [comment]
@gdha @pcahyna 
Rear 3.0 milestone is fine with me.
I would prefer to have it merged as is
over not having this improvement in ReaR 3.0
because it is already a good improvement
regardless of some still existing issues.

**Comment by jsmeix** [comment]
@pcahyna @rear/contributors
I would like to merge it this week on Thursday afternoon
(or sooner if @pcahyna agrees) unless there are objections.

**Comment by jsmeix** [comment]
@pcahyna 
I would like to merge it today afternoon, cf.
https://github.com/rear/rear/pull/3125#issuecomment-2172462514
unless there are objections.
